Make ToggleBase a subclass of ButtonBase

matthewcoufal opened the pull request:
This PR adjusts `ToggleBase` to be a subclass of `ButtonBase` to address the issue https://github.com/mit-cml/appinventor-sources/issues/1824.

AppInventorWorkerBee commented:
Can one of the admins verify this patch?

ewpatton commented:
@matthewcoufal We also need to rebase this onto `ucr` since it will be changing the API exposed by the companion app.

matthewcoufal commented:
@ewpatton I addressed your comments with my most recent commit. Let me know if there is anything else you want me to do for this PR - thanks for the help!

ewpatton commented:
@matthewcoufal You also need to rebase your branch. The following command should do what you need:

```
git rebase upstream/master --onto upstream/ucr
```

When you push, you will need to use a force-push since your branch will have the new base.

matthewcoufal commented:
@ewpatton Rebased and force-pushed. Is there anything else that needs to be done?

ewpatton commented:
@AppInventorWorkerBee ok to test

matthewcoufal commented on appinventor/components/src/com/google/appinventor/components/runtime/ToggleBase.java:
```diff
@@ -18,33 +18,16 @@
 import android.view.View.OnFocusChangeListener;
 import android.widget.CompoundButton;
 import android.widget.CompoundButton.OnCheckedChangeListener;
+import javafx.scene.text.TextAlignment;
```
I think this was an unintentional change from my editor. Will remove this now.

ewpatton replied:
@matthewcoufal I'm still seeing this in the PR. Please remove.

matthewcoufal commented on appinventor/components/src/com/google/appinventor/components/runtime/ToggleBase.java:
```diff
@@ -73,45 +56,74 @@ protected void initToggle() {
 
   }
 
+  @Override
```
@ewpatton I think it needs to be overridden here because `ToggleBase` now extends `ButtonBase`, which has a `click` event. Thoughts?

matthewcoufal commented on appinventor/components/src/com/google/appinventor/components/runtime/ToggleBase.java:
```diff
@@ -73,45 +56,74 @@ protected void initToggle() {
+      category = PropertyCategory.APPEARANCE,
+      description = "Left, center, or right.",
+      userVisible = false)
+  public int TextAlignment() {
```
@ewpatton I tried this (adjusted this function to return `textAlignment`, a `protected` variable inherited from `ButtonBase`), and when I changed the text alignment there seemed to be no effect (for both the `CheckBox` and `Switch` components).

matthewcoufal commented:
@ewpatton Is this good to go, or is there anything else that needs to be changed?

ewpatton commented on appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java:
```diff
@@ -1386,8 +1392,8 @@ private void processProperties(ComponentInfo componentInfo,
               throw new RuntimeException("Inconsistent types " + priorProperty.type +
-                                         " and " + newProperty.type + " for property " +
-                                         propertyName + " in component " + componentInfo.name);
+                                        " and " + newProperty.type + " for property " +
```
Looks like your editor played games with the indentation in this file. Correct the lines with whitespace only changes to reflect what's in the base branch.

ewpatton commented:
I'm also seeing a lot of whitespace preceding the ToggleBase components

![Screenshot_1575496658](https://user-images.githubusercontent.com/1331253/70185159-4626de00-16b7-11ea-94df-df1faf8b852f.png)

ewpatton commented:
@matthewcoufal TextAlignment also didn't work for me for the Checkbox and Switch. You may want to just `@Override` that so it doesn't appear in the designer unless it would be straightforward to get it working.

matthewcoufal commented:
@ewpatton Added some changes. Let me know if this does not fully address what you had requested.
